p2p/ipld/read.go

```diff
@@ -224,53 +222,40 @@ func GetLeafData(
 	rootCid cid.Cid,
 	leafIndex uint32,
 	totalLeafs uint32, // this corresponds to the extended square width
-	api coreiface.CoreAPI,
+	dag format.NodeGetter,
 ) ([]byte, error) {
-	// calculate the path to the leaf
-	leafPath, err := leafPath(leafIndex, totalLeafs)
+	nd, err := GetLeaf(ctx, dag, rootCid, leafIndex, totalLeafs)
 	if err != nil {
 		return nil, err
 	}
 
-	// use the root cid and the leafPath to create an ipld path
-	p := path.Join(path.IpldPath(rootCid), leafPath...)
+	return nd.RawData()[1:], nil
+}
 
-	// resolve the path
-	node, err := api.ResolveNode(ctx, p)
+// GetLeafData fetches and returns the raw leaf.
+// It walks down the IPLD NMT tree until it finds the requested one.
+func GetLeaf(ctx context.Context, dag format.NodeGetter, root cid.Cid, leaf, total uint32) (format.Node, error) {
+	// request the node
+	nd, err := dag.Get(ctx, root)
```

Comment:
the new `GetLeaf` is much simpler! I like it a lot :+1:

Reply:
@liamsi do we need to update ADR002 with this slight api change?

Reply:
Yeah, it would be good if the ADR would match the implementation's APIs.

Reply:
I think so, but also I am going to propose an elegant PR soon. If it's get accepted, updating this will make more sense